### PR TITLE
feat(framework) Limit TaskRes TTL when saving it.

### PR DESCRIPTION
**Pull request by mohammadnaseri**
### Description
This PR limits TaskRes TTL with respect to `TaskRes.created_at`, `TaskIns.created_at`, and `TaskIns.ttl`. The condition is:
`TaskIns.created_at + TaskIns.ttl ≥ TaskRes.created_at + TaskRes.ttl`. 
Please merge it after #3609

### CODE REVIEW

File: src/py/flwr/server/superlink/state/in_memory_state.py [hunk 1]
```diff
@@ -117,6 +117,35 @@ def store_task_res(self, task_res: TaskRes) -> Optional[UUID]:
...
+                task_res.task.ttl
+                > task_ins.task.created_at
+                + task_ins.task.ttl
+                - task_res.task.created_at
+            ):
+                task_res.task.ttl = (
+                    task_ins.task.created_at
+                    + task_ins.task.ttl
+                    - task_res.task.created_at
+                )
```
Comment (panh99) [inline src/py/flwr/server/superlink/state/in_memory_state.py:147]: I think we should log a warning/info if the `SuperLink` receives a `TaskRes` with larger than expected TTL and updates it.

File: src/py/flwr/server/superlink/state/sqlite_state.py [hunk 2]
```diff
@@ -347,7 +347,29 @@ def store_task_res(self, task_res: TaskRes) -> Optional[UUID]:
...
+        # expiration time of the TaskIns it replies to.
+        # Condition: TaskIns.created_at + TaskIns.ttl ≥
+        #            TaskRes.created_at + TaskRes.ttl
+        if (
+            task_res.task.ttl
+            > task_ins["created_at"] + task_ins["ttl"] - task_res.task.created_at
+        ):
+            task_res.task.ttl = (
+                task_ins["created_at"] + task_ins["ttl"] - task_res.task.created_at
+            )
```
Comment (panh99) [inline src/py/flwr/server/superlink/state/sqlite_state.py:370]: Same as above.

File: src/py/flwr/server/superlink/state/in_memory_state.py [hunk 3]
```diff
@@ -134,6 +134,23 @@ def store_task_res(self, task_res: TaskRes) -> Optional[UUID]:
...
+            )
+            if task_res.task.ttl > max_allowed_ttl:
+                log(
+                    WARNING,
+                    "Received TaskRes with TTL %s exceeding the allowed maximum TTL %s."
+                    "Updating TTL to the allowed maximum.",
+                    task_res.task.ttl,
+                    max_allowed_ttl,
+                )
+                task_res.task.ttl = max_allowed_ttl
```
Comment (danieljanes) [inline src/py/flwr/server/superlink/state/in_memory_state.py:152]: Instead of updating the TTL, we should fail. It the provided TTL is wrong here, it's likely due to a client-side bug that should be addressed.

File: src/py/flwr/server/superlink/state/sqlite_state.py [hunk 4]
```diff
@@ -383,6 +383,23 @@ def store_task_res(self, task_res: TaskRes) -> Optional[UUID]:
...
+        )
+        if task_res.task.ttl > max_allowed_ttl:
+            log(
+                WARNING,
+                "Received TaskRes with TTL %s exceeding the allowed maximum TTL %s."
+                "Updating TTL to the allowed maximum.",
+                task_res.task.ttl,
+                max_allowed_ttl,
+            )
+            task_res.task.ttl = max_allowed_ttl
```
Comment (danieljanes) [inline src/py/flwr/server/superlink/state/sqlite_state.py:401]: Instead of updating the TTL, we should fail. It the provided TTL is wrong here, it's likely due to a client-side bug that should be addressed.

File: src/py/flwr/common/message_test.py [hunk 5]
```diff
@@ -202,3 +202,36 @@ def test_repr(cls: type, kwargs: dict[str, Any]) -> None:
...
+        # Case where the reply_ttl is within the allowed range
+        (create_message_with_content, 20, 10, 10),
+        (create_message_with_error, 20, 10, 10),
+    ],
+)
+def test_reply_ttl_limitation(
+    message_creation_fn: Callable[[float], Message],
+    initial_ttl: float,
+    reply_ttl: float,
+    expected_max_ttl: float,
```
Comment (panh99) [inline src/py/flwr/common/message_test.py:222]: Based on your code, this should be `expected_reply_ttl`, no?

File: src/py/flwr/common/message_test.py [hunk 6]
```diff
@@ -202,3 +202,36 @@ def test_repr(cls: type, kwargs: dict[str, Any]) -> None:
...
+    time.sleep(0.1)
+
+    if message.has_error():
+        dummy_error = Error(code=0, reason="test error")
+        reply_message = message.create_error_reply(dummy_error, ttl=reply_ttl)
+    else:
+        reply_message = message.create_reply(content=RecordSet(), ttl=reply_ttl)
+
+    assert (
+        reply_message.metadata.ttl <= expected_max_ttl
```
Comment (panh99) [inline src/py/flwr/common/message_test.py:236]: I think we should only limit the TTL if it exceeds the allowed maximum. We should also prevent our code make the actual TTL unnecessarily small.
```suggestion
        abs(reply_message.metadata.ttl - expected_reply_ttl) <= 1e-9
```

File: src/py/flwr/server/superlink/state/in_memory_state.py [hunk 7]
```diff
@@ -134,6 +134,26 @@ def store_task_res(self, task_res: TaskRes) -> Optional[UUID]:
                 )
                 return None
 
+            # Fail if the TaskRes TTL exceeds the
+            # expiration time of the TaskIns it replies to.
+            # Condition: TaskIns.created_at + TaskIns.ttl ≥
+            #            TaskRes.created_at + TaskRes.ttl
+            # A small tolerance is introduced to account
+            # for floating-point precision issues.
+            tolerance = 1e-2
```
Comment (panh99) [inline src/py/flwr/server/superlink/state/in_memory_state.py:143]: It's better to make it a constant.

File: src/py/flwr/server/superlink/state/sqlite_state.py [hunk 8]
```diff
@@ -383,6 +383,25 @@ def store_task_res(self, task_res: TaskRes) -> Optional[UUID]:
             )
             return None
 
+        # Fail if the TaskRes TTL exceeds the
+        # expiration time of the TaskIns it replies to.
+        # Condition: TaskIns.created_at + TaskIns.ttl ≥
+        #            TaskRes.created_at + TaskRes.ttl
+        # A small tolerance is introduced to account
+        # for floating-point precision issues.
+        tolerance = 1e-2
```
Comment (panh99) [inline src/py/flwr/server/superlink/state/sqlite_state.py:392]: Same as above.

**Review by panh99 — CHANGES_REQUESTED** [review]

File: src/py/flwr/server/superlink/state/state_test.py [hunk 9]
```diff
@@ -707,6 +707,42 @@ def test_store_task_res_task_ins_expired(self) -> None:
         # Assert
         assert result is None
 
+    def test_store_task_res_limit_ttl(self) -> None:
```
Comment (panh99) [inline src/py/flwr/server/superlink/state/state_test.py:710]: Should we use `parameterized` to test multiple pairs of `created_at` and `ttl`? This function only test the positive cases, right? I think we should also test negative cases.

**Review by panh99 — DISMISSED** [review]

File: src/py/flwr/common/message.py [hunk 10]
```diff
@@ -309,6 +303,23 @@ def create_error_reply(self, error: Error, ttl: float | None = None) -> Message:
...
+        # Condition: msg.metadata.ttl + msg.metadata.created_at ≥
+        #            reply_msg.metadata.ttl + reply_msg.metadata.created_at must
+        max_allowed_ttl = (
+            self.metadata.created_at + self.metadata.ttl - message.metadata.created_at
+        )
+        if message.metadata.ttl > max_allowed_ttl:
+            log(
+                WARNING,
+                "The reply TTL %s exceeding the allowed maximum TTL %s."
+                "Updating TTL to the allowed maximum.",
```
[the remark this answers is not in the document]
Comment (mohammadnaseri) [reply src/py/flwr/common/message.py]: let's not use a unit yet until we decide about it later.

[the remark this answers is not in the document]
Comment (panh99) [reply src/py/flwr/common/message.py]: This is a user-facing log message, so I believe we should include the unit. Even if we update how we store these values in our codebase, we will still display TTL in seconds for our users.

File: src/py/flwr/server/superlink/state/state_test.py (hunk 9 above)
Comment (danieljanes) [reply src/py/flwr/server/superlink/state/state_test.py:710]: Yes, good idea

File: src/py/flwr/server/superlink/state/in_memory_state.py [hunk 11]
```diff
@@ -134,6 +134,28 @@ def store_task_res(self, task_res: TaskRes) -> Optional[UUID]:
...
+                and (task_res.task.ttl - max_allowed_ttl) > TTL_TOLERANCE
+            ):
+                log(
+                    ERROR,
+                    "Received TaskRes with TTL %.2f "
+                    "exceeding the allowed maximum TTL %.2f.",
+                    task_res.task.ttl,
+                    max_allowed_ttl,
+                )
+                return None
```
Comment (danieljanes) [inline src/py/flwr/server/superlink/state/in_memory_state.py:157]: This block of code and the one in `sqlite_state` should live in a shared function to reduce code duplication.

File: src/py/flwr/common/message.py [hunk 12]
```diff
@@ -309,6 +304,26 @@ def create_error_reply(self, error: Error, ttl: float | None = None) -> Message:
...
+        if message.metadata.ttl - max_allowed_ttl > TTL_TOLERANCE:
+            log(
+                WARNING,
+                "The reply TTL of %.2f seconds exceeded the "
+                "allowed maximum of %.2f seconds."
+                "The TTL has been updated to the allowed maximum.",
+                ttl,
+                max_allowed_ttl,
+            )
+            message.metadata.ttl = max_allowed_ttl
```
Comment (danieljanes) [inline src/py/flwr/common/message.py:325]: This block of code and the one below should live in a shared function to reduce code duplication.

File: src/py/flwr/common/message_test.py [hunk 13]
```diff
@@ -202,3 +203,36 @@ def test_repr(cls: type, kwargs: dict[str, Any]) -> None:
...
+    time.sleep(0.1)
+
+    if message.has_error():
+        dummy_error = Error(code=0, reason="test error")
+        reply_message = message.create_error_reply(dummy_error, ttl=reply_ttl)
+    else:
+        reply_message = message.create_reply(content=RecordSet(), ttl=reply_ttl)
+
+    assert (
+        reply_message.metadata.ttl - expected_reply_ttl <= TTL_TOLERANCE
```
Comment (panh99) [inline src/py/flwr/common/message_test.py:237]:
```suggestion
        abs(reply_message.metadata.ttl - expected_reply_ttl) <= TTL_TOLERANCE
```

Comment (mohammadnaseri) [reply src/py/flwr/common/message_test.py:237]: @panh99 Did you test the code with this change? It makes the test failing.

File: src/py/flwr/server/superlink/state/state_test.py (hunk 9 above)
Comment (mohammadnaseri) [reply src/py/flwr/server/superlink/state/state_test.py:710]: I added more test cases, but after a lot of effort, I realized that using `parameterized` tests and `time.time()` makes things flaky. That is why I avoided using it.

File: src/py/flwr/common/message.py [hunk 14]
```diff
@@ -370,6 +362,30 @@ def __repr__(self) -> str:
         )
         return f"{self.__class__.__qualname__}({view})"
 
+    def _limit_task_res_ttl(self, message: Message) -> None:
+        """Limit the TaskRes TTL to not exceed the expiration time of the TaskIns it
+        replies to.
+
+        Parameters
+        ----------
+            message: The message to which the TaskRes is replying.
```
Comment (danieljanes) [inline src/py/flwr/common/message.py:371]:
```suggestion
        message : Message
            The message to which the TaskRes is replying.
```

File: src/py/flwr/server/superlink/state/in_memory_state.py [hunk 15]
```diff
@@ -134,6 +138,27 @@ def store_task_res(self, task_res: TaskRes) -> Optional[UUID]:
...
+            if task_res.task.ttl and (
+                task_res.task.ttl - max_allowed_ttl > MESSAGE_TTL_TOLERANCE
+            ):
+                # log(
+                #     ERROR,
+                #     "Received TaskRes with TTL %.2f "
+                #     "exceeding the allowed maximum TTL %.2f.",
+                #     task_res.task.ttl,
+                #     max_allowed_ttl,
+                # )
```
[the remark this answers is not in the document]
Comment (mohammadnaseri) [reply src/py/flwr/server/superlink/state/in_memory_state.py]: that is not supposed to be commented out. I just correct it.

**Review by danieljanes — DISMISSED** [review]

**Review by danieljanes — APPROVED** [review]
